### PR TITLE
Frontend libraries: reduce dependency on CDN #7180

### LINKED ISSUE #7180: Frontend libraries: reduce dependency on CDN
One of the frontend libraries, https://unpkg.com/guillotine@1.3.1/js/jquery.guillotine.min.js, [can't be accessed from the build environment](https://ci.appveyor.com/project/whipermr5/teammates/build/1.0.32#L198), causing our recent streak of AppVeyor build failures. It looks like a bug in unpkg: https://github.com/unpkg/express-unpkg/issues/70

This bug has the potential to affect all our users based in the same geographical location as the AppVeyor servers (seems to be U.S. based on their build environment IP). The effect for now is small (unable to edit profile pictures using the web UI), but it does raise the question of whether we've become over-reliant on CDNs (or rather, a single CDN) since #6346 (#6373).

Possible fix: failover to local copy?

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #7180, or so I hope.

**Outline of Solution**

Pre-works done (up to 835d70c):
- Moved all scripts to bottom of the page; these were the missed cases in #6105
- Moved all CSS definitions to `teammatesCommon.css` to finally enforce our "one-CSS-file" rule. One of the CSS files contains entirely unused rules.

Works done (everything after):
Made full use of Webpack to bundle the dependencies.
- `libs-common` for the libraries used by all pages: jQuery, jQuery-ui, all jQuery plugins (just to make it easier for now; can consider separating later on), Bootstrap and its CSS files, bootbox, and `teammatesCommon.css`.
- `libs-tinymce`, `libs-moment`, `libs-map`, `libs-testing`, `libs-qunit` as necessary.

**Advantages**

- All the advantages of I have mentioned in the issue thread are achieved. Arguably, this is the most maintainable way of managing our third-party front-end libraries that does not necessitate any reliance on any CDN.
- Front-end performance boost: less library files to be loaded which means less HTTP requests (in some cases the total file size is even less) and minified CSS.

**Caveat**

I need to warn that this PR, at its current state (latest commit be23e95), has a number of caveats:
- Some of the ways to bundle certain libraries may be improper/"hacks", partly because I am still trying to understand Webpack and some were solutions that I came up on the spot while learning.
- While I can give a "both-CIs-green" guarantee, I cannot guarantee that everything still works the way they were before; some things are not covered in tests, some are corner cases, and some are simply not meant to be tested because they are only for visual eyeballing.
- This last caveat will not be gone as long as this PR exists: as I mentioned in the issue thread, adding a new library (if necessary) will not be as simple as before; need to make sure that the new library works well with Webpack. It does not help that Webpack has a rather steep learning curve.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -36,10 +36,10 @@
     "build:src": "webpack --config webpack-main.config.js",
     "build:test": "webpack --config webpack-test.config.js",
     "build": "npm run build:libs && npm run build:src && npm run build:test",
-    "lint:js": "eslint src/main/webapp/dev -c static-analysis/teammates-eslint.yml --ext es6",
-    "lint:json": "eslint src/main/webapp/data/*.json src/main/resources/*.json src/test/resources/data/*.json -c static-analysis/teammates-eslint-json.yml",
-    "lint:css": "stylelint src/main/webapp/stylesheets/*.css --config static-analysis/teammates-stylelint.yml",
-    "lint:spaces": "lintspaces -n -t -d spaces -l 1 -. \"src/**/*.jsp\" \"src/**/*.tag\" \"src/main/**/*.html\" \"src/**/*.xml\" \"src/**/*.json\" \"src/**/*.css\" \"src/**/*.java\" \"src/**/*.es6\" \"src/**/*.properties\" \"*.yml\" \"*.gradle\" \"static-analysis/*.*ml\"",
+    "lint:js": "eslint *.js src/main/webapp/dev/**/*.es6 -c static-analysis/teammates-eslint.yml",
```
Comment (whipermr5) [inline package.json]: Does this mean it'll lint all the generated `.js` files too (if we run `npm run lint` after `npm run build`?

File: webpack-libs.config.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        ],
+    },
+    stats: 'errors-only',
+    plugins: [
+        new webpack.optimize.UglifyJsPlugin({
+            output: {
+                ascii_only: true,
+            },
+        }),
+        new webpack.IgnorePlugin(/^\.(\/|\\)locale$/, /moment$/),
```
Comment (whipermr5) [inline webpack-libs.config.js]: I'm guessing this is because `moment.js` is incompatible with `webpack`?

File: package.json (hunk 1 above)
Comment (wkurniawan07) [inline package.json]: No, that's not how the inclusion pattern works. This will only target the JS files at the root, which will be the Webpack config files.

File: package.json [hunk 3]
```diff
@@ -7,28 +16,16 @@
...
     "webpack": "2.6.1"
   },
-  "babel": {
-    "presets": [
-      [
-        "env",
-        {
-          "targets": {
-            "browsers": [
-              "ie >= 9"
```
Comment (wkurniawan07) [inline package.json]: This should have been done in #7522. As far as I understand, this is only used by Babel CLI (Babel loader has its own separate config) and thus is no longer relevant now -- I could be wrong here.

File: webpack-libs.config.js (hunk 2 above)
Comment (wkurniawan07) [inline webpack-libs.config.js]: https://webpack.js.org/plugins/ignore-plugin/#ignore-moment-locales
I don't see why the locales are needed and they bloat the library size.

File: webpack-libs.config.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    if (fileName.endsWith('.es6')) {
+        const fileNameWithoutExt = fileName.replace('.es6', '');
+        const filesToBundle = [
+            `${LIBS_JS_FOLDER}/${fileName}`,
+        ];
+        entry[fileNameWithoutExt] = filesToBundle;
+    }
+});
+
+const jquery = {
```
Comment (thenaesh) [inline webpack-libs.config.js]: Should change this name? Looks like it contains more than just jQuery.

Comment (whipermr5) [inline webpack-libs.config.js]: Maybe `core-ui`?

**Comment by whipermr5** [comment]
@wkurniawan07 Any updates here? :)

**Comment by wkurniawan07** [comment]
Haven't got the time to work on this for the past few days due to distraction from Java 8 😅 will try and push something by tomorrow.

**Comment by wkurniawan07** [comment]
Closing this as agreed during discussion with @whipermr5. The branch containing the changes here will be kept intact in case this issue wants to be worked again the future.
